stumpy/mmotifs.py

```diff
@@ -0,0 +1,210 @@
+        equivalent set in the `@core.non_normalized` function decorator.
+
+    p : float, default 2.0
+        The p-norm to apply for computing the Minkowski distance. This parameter is
+        ignored when `normalize == False`.
+
+    Returns
+    -------
+    motif_distances: numpy.ndarray
+        The distances corresponding to a set of subsequence matches for each motif.
+
+    motif_indices: numpy.ndarray
+        The indices corresponding to a set of subsequences matches for each motif.
+
+    motif_subspaces: numpy.ndarray
+        A numpy.ndarray consisting of arrays that contain the `k`-dimensional
+        subspace for each motif.
+
+    motif_mdls: numpy.ndarray
+        A numpy.ndarray consisting of arrays that contain the mdl results for
+        finding the dimension of each motif
+    """
+    # Convert dataframe to array if necessary
+    T = core._preprocess(T)
+
+    if max_motifs < 1:  # pragma: no cover
+        logger.warning(
+            "The maximum number of motifs, `max_motifs`, "
+            "must be greater than or equal to 1"
+        )
+        logger.warning("`max_motifs` has been set to `1`")
+        max_motifs = 1
+
+    # Calculate subsequence length
+    m = T.shape[-1] - P.shape[-1] + 1
+
+    # Calculate exclusion zone
+    excl_zone = int(np.ceil(m / config.STUMPY_EXCL_ZONE_DENOM))
+
+    # Set default max_matches value
+    if max_matches is None:
+        max_matches = np.inf
+
+    # Calculate default cutoff values
+    if cutoffs is None:
+        cutoffs = []
+        P_copy = P.copy().astype(np.float64)
+        P_copy[np.isinf(P_copy)] = np.nan
+        for i in range(P_copy.shape[0]):
+            cutoffs.append(
+                np.nanmax(
+                    [
+                        np.nanmean(P_copy[i, :] - 2.0 * np.nanstd(P_copy[i, :])),
+                        np.nanmin(P_copy[i, :]),
+                    ]
+                )
+            )
+    elif isinstance(cutoffs, int) or isinstance(cutoffs, float):
+        # cutoffs is a single value -> apply value to each MP dimension
+        cutoffs_tmp = cutoffs
+        cutoffs = []
+        cutoffs.extend(cutoffs_tmp for _ in range(T.shape[0]))
+
+    # Precompute rolling means and standard deviations
+    T, M_T, Σ_T = core.preprocess(T, m)
+
+    # Allocate memory for returns
+    motif_distances = []
+    motif_indices = []
+    motif_subspaces = []
+    motif_mdls = []
+
+    # Identify the first multidimensional motif as candidate motif and get its
+    # nearest neighbor
+    candidate_idx = np.argmin(P, axis=1)
+    nn_idx = I[np.arange(len(candidate_idx)), candidate_idx]
+
+    # Iterate over each motif
+    while len(motif_distances) < max_motifs:
+
+        # Choose dimension k using MDL
+        mdls, subspaces = mdl(T, m, candidate_idx, nn_idx)
+        k = np.argmin(mdls)
+        motif_mdls.append(mdls)
+
+        # Choose cutoff value for the `k`-th dimension
+        cutoff = cutoffs[k]

```

Comment:
Instead of using `cutoff`, I think it is better to simply use `cutoffs[k]` everywhere. For me, seeing `cutoffs[k]` reminds me that it is a sub-dimension